neqo-http3-server/src/main.rs

```diff
@@ -207,234 +209,299 @@ fn process(
-                        if let Some((ref mut server, svr_timeout)) =
-                            servers.get_mut(&socket.local_addr().unwrap())
-                        {
-                            process(&mut **server, svr_timeout, inx, None, &mut timer, socket);
-                        }
-                    }
+        for (i, host) in self.hosts.iter().enumerate() {
+            let socket = match UdpSocket::bind(&host) {
+                Err(err) => {
+                    eprintln!("Unable to bind UDP socket: {}", err);
+                    return Err(err);
                 }
-            } else if let Some(socket) = sockets.get_mut(event.token().0) {
-                let local_addr = hosts[event.token().0];
+                Ok(s) => s,
+            };
 
-                if !event.readiness().is_readable() {
-                    continue;
+            let local_addr = match socket.local_addr() {
+                Err(err) => {
+                    eprintln!("Socket local address not bound: {}", err);
+                    return Err(err);
                 }
+                Ok(s) => s,
+            };
+
+            let res = socket.only_v6();
+            let also_v4 = if res.is_ok() && !res.unwrap() {
+                " as well as V4"
+            } else {
+                ""
+            };
+            println!(
+                "Server waiting for connection on: {:?}{}",
+                local_addr, also_v4
+            );
+
+            self.poll.register(
+                &socket,
+                Token(i),
+                Ready::readable() | Ready::writable(),
+                PollOpt::edge(),
+            )?;
+
+            self.sockets.push(socket);
+            self.servers
+                .insert(local_addr, (create_server(&self.args), None));
+        }
 
-                loop {
-                    let (sz, remote_addr) = match socket.recv_from(&mut buf[..]) {
-                        Err(ref err) if err.kind() == io::ErrorKind::WouldBlock => break,
-                        Err(err) => {
-                            eprintln!("UDP recv error: {:?}", err);
-                            exit(1);
-                        }
-                        Ok(res) => res,
-                    };
+        self.poll
+            .register(&self.timer, TIMER_TOKEN, Ready::readable(), PollOpt::edge())?;
 
-                    if sz == buf.len() {
-                        eprintln!("Might have received more than {} bytes", buf.len());
-                    }
+        Ok(())
+    }
 
-                    if sz == 0 {
-                        eprintln!("zero length datagram received?");
-                    } else if let Some((ref mut server, svr_timeout)) =
-                        servers.get_mut(&socket.local_addr().unwrap())
-                    {
-                        process(
+    fn process_and_events(&mut self, inx: usize, read_socket: bool) -> Result<(), io::Error> {
```

Comment:
The name here is strange.  I think that this is process_datagrams_and_events, but that's pretty lengthy.